From a6e957a93f43546a13891c29aff5f4a0115262bc Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Mon, 07 Sep 2020 12:16:09 +0300
Subject: [PATCH] [SW-19543] gaudi: validate security strap

Add validation to security strap value. The reason is Gaudi boards with
HW rev lower then E which has undetermined security strap value.
Validate this value by checking if OTP is valid.
Ignore strap if otherwise.

Change-Id: I2dc3a492582cdfa543c92dd9e61b24ffbd7cd5cf
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index dd7acb8..9e5e5a7 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -19,6 +19,7 @@
 #include <string.h>
 #include <efuse.h>
 #include <gaudi/gaudi_security.h>
+#include <spi.h>
 
 struct preboot_security_efuse_cache {
 	bool is_invalid;
@@ -28,6 +29,32 @@
 static struct preboot_security_efuse_cache preboot_security_fuse_cache = {
 	.is_invalid = true,
 };
+
+struct preboot_security_otp_cache {
+	bool is_invalid;
+	struct hl_sec_otp otp;
+};
+
+struct preboot_security_otp_cache preboot_security_otp_cache = {
+	.is_invalid = true,
+};
+
+static const struct hl_sec_otp *preboot_security_get_otp(void)
+{
+	struct hl_sec_otp *otp = &preboot_security_otp_cache.otp;
+
+	if (preboot_security_otp_cache.is_invalid) {
+		spi_copy_otp(otp, OTP_SECURITY_OFFSET,
+			     sizeof(struct hl_sec_otp));
+
+		if (otp->magic_number != OTP_SECURITY_MAGIC)
+			return NULL;
+
+		preboot_security_otp_cache.is_invalid = false;
+	}
+
+	return otp;
+}
 
 static const struct hl_sec_efuse *preboot_security_get_fuse(void)
 {
@@ -62,7 +89,21 @@
 
 	boot_strap = regmap_readl(mmPSOC_GLOBAL_CONF_BOOT_STRAP_PINS);
 
-	return !(boot_strap & SECURITY_ENABLE_SW);
+	if (boot_strap & SECURITY_ENABLE_SW)
+		return false;
+
+	/* Check if OTP is valid.
+	 * This is done due to boards with HW rev lower then E which
+	 * has undetermined security strap value.
+	 * In case that security strap is 0 (on) and OTP is invalid, ignore the
+	 * strap.
+	 */
+	if (!preboot_security_get_otp()) {
+		preboot_puts("*** Security strap is ON while OTP is invalid - ignoring strap\n");
+		return false;
+	}
+
+	return true;
 }
 
 /* Security is enabled in the following condition:
